src/logging.h

```diff
@@ -155,12 +155,10 @@ static inline void LogPrintf(const char* fmt, const Args&... args)
     }
 }
 
-template <typename... Args>
-static inline void LogPrint(const BCLog::LogFlags& category, const Args&... args)
-{
-    if (LogAcceptCategory((category))) {
-        LogPrintf(args...);
-    }
-}
+#define LogPrint(category, ...) do { \
```

Comment:
Just a suggestion, but it might be good to leave existing LogPrint function alone, and instead just add a new `LOG_CATEGORY(category, format, ...)` macro. This would avoid issue the Jeremy raised about side effects in existing code, and also be nicer in my opinion because uppercase naming would be more consistent with other macros.

Reply:
My feeling is that providing two ways to log where one is subtly different than the other adds unnecessary cognitive load to readers and reviewers. Will address Jeremey's comment in a follow-up.

Reply:
> My feeling is that providing two ways to log where one is subtly different than the other adds unnecessary cognitive load to readers and reviewers. Will address Jeremey's comment in a follow-up.

If the concern is cognitive load, I'd think the current:

- `LogPrintf` - log unconditionally
- `LogPrint` - log for category

is more confusing than what I'm proposing:

- `LogPrintf` - log unconditionally
- `LOG_CATEGORY` - log for category

But if it would take a long time to deprecate `LogPrint` (not sure why it would), then I agree inconsistency would not look nice in the interim.

I don't feel strongly any way about this, so please ignore this suggestion if doesn't suit you.

Reply:
Ah, I thought you were suggesting `LogPrint`, `LogPrintf`, and `LOG_CATEGORY`.

I'm not completely opposed to changing call sites to `LOG_CATEGORY` (and checking for side effects in the process). But I would like to be sure such a change would be welcome before putting in the work. :)

Reply:
> I would like to be sure such a change would be welcome before putting in the work. :)

To be clear, suggestion is _not_ to do more work. My suggestion is:

1. Add your new macro with a name like `LOG_CATEGORY`, to be clear that it is a macro and not something that evaluated like a normal function.
2. Leave `LogPrint` alone, mark it deprecated and add a note like "LOG_CATEGORY is preferred over LogPrint in new code as a more performant alternative"

Separately after that, I do not think it would take a lot of work to replace instances of `LogPrint` and remove it. We could start with a simple [scripted-diff](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#scripted-diffs) to replace simple variable references that obviously don't have side effects.

```sh
git grep -l LogPrint | xargs sed -i 's/LogPrint\((BCLog::[A-Z]\+, "[^"]\+"\(, [A-Za-z_]\+\)*)\)/LOG_CATEGORY\1/g'
```

As for organizing these changes, I might consider closing this PR, and instead adding the new macro directly where you intend to use it in #16688.

Reply:
Understood. I would be fine with that assuming there is agreement that `LogPrint` should be deprecated in favor of the `LOG_CATEGORY` macro.